parachain/src/xcm.rs

```diff
@@ -0,0 +1,238 @@
+#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+pub enum VersionedMultiLocation {
+	V0(v0::MultiLocation),
+}
+
+/// A versioned multi-asset, an identifier for an asset within a consensus system.
+#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+pub enum VersionedMultiAsset {
+	V0(v0::MultiAsset),
+}
+
+pub mod v0 {
+	use super::*;
+
+	/// Basically just the XCM (more general) version of `ParachainDispatchOrigin`.
+	#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+	pub enum MultiOrigin {
+		/// Origin should just be the native origin for the sender. For Cumulus/Frame chains this is
+		/// the `Parachain` origin.
+		Native,
+		/// Origin should just be the standard account-based origin with the sovereign account of
+		/// the sender. For Cumulus/Frame chains, this is the `Signed` origin.
+		SovereignAccount,
+		/// Origin should be the super-user. For Cumulus/Frame chains, this is the `Root` origin.
+		/// This will not usually be an available option.
+		Superuser,
+	}
+
+	impl From<ParachainDispatchOrigin> for MultiOrigin {
+		fn from(o: ParachainDispatchOrigin) -> Self {
+			match o {
+				ParachainDispatchOrigin::Parachain => MultiOrigin::Native,
+				ParachainDispatchOrigin::Signed => MultiOrigin::SovereignAccount,
+				ParachainDispatchOrigin::Root => MultiOrigin::Superuser,
+			}
+		}
+	}
+
+	impl From<MultiOrigin> for ParachainDispatchOrigin {
+		fn from(o: MultiOrigin) -> Self {
+			match o {
+				MultiOrigin::Native => ParachainDispatchOrigin::Parachain,
+				MultiOrigin::SovereignAccount => ParachainDispatchOrigin::Signed,
+				MultiOrigin::Superuser => ParachainDispatchOrigin::Root,
+			}
+		}
+	}
+
+	#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+	pub enum MultiNetwork {
+		Wildcard,
+		Identified(Vec<u8>),
+	}
+
+	#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+	pub enum MultiLocation {
+		Null,
+		X1(Junction),
+		X2(Junction, Junction),
+		X3(Junction, Junction, Junction),
+		X4(Junction, Junction, Junction, Junction),
+	}
+
+	#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+	pub enum Junction {
+		Parent,
+		Parachain { #[codec(compact)] id: u32 },
+		OpaqueRemark(Vec<u8>),
+		AccountId32 { network: MultiNetwork, id: [u8; 32] },
+		AccountIndex64 { network: MultiNetwork, #[codec(compact)] index: u64 },
+		AccountKey20 { network: MultiNetwork, key: [u8; 20] },
+		/// An instanced Pallet on a Frame-based chain.
+		PalletInstance { id: u8 },
+		/// A nondescript index within the context location.
+		GeneralIndex { #[codec(compact)] id: u128 },
+		/// A nondescript datum acting as a key within the context location.
+		GeneralKey(Vec<u8>),
+	}
+
+	impl From<Junction> for MultiLocation {
+		fn from(x: Junction) -> Self {
+			MultiLocation::X1(x)
+		}
+	}
+
+	#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+	pub enum AssetInstance {
+		Undefined,
+		Index8(u8),
+		Index16 { #[codec(compact)] id: u16 },
```

Comment:
Not sure if we want to use compact encoding for u16, it can only save up to one byte and can use two more bytes in some cases.